### PR TITLE
kernel: make HAVE_WORK a member variable

### PR DESCRIPTION
**Pull request by bradjc**
`HAVE_WORK` is currently a process.rs global static variable. This is probably not what we really want, as it requires `unsafe` to use. This creates a `Kernel` struct and stores the number of callbacks waiting as a member variable in that struct. In the process, I counted 5 uses of `unsafe` that are now gone.

Now, before HAVE_WORK was a `VolatileCell`. I'm not sure why that is needed, so I made it a `NumCell` instead (essentially just a `Cell`). If someone knows why it must be volatile then it can be changed back.

Also, this moves the schedule function to a be a class function of `Process`, where it probably should have been anyway.

### Testing Strategy

This pull request was tested by running hail on hail.

### TODO or Help Wanted

n/a

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
I want to note that this PR is in service to tracking issue #1043.

**Comment by alevy** [comment]
> Now, before HAVE_WORK was a VolatileCell

This seems to be important for the compiler to elide writes to static mutables that are not read within the same crate. Or at least was. So it was a hack, and I don't think we should need it if it's a member field.

**Comment by alevy** [comment]
These changes have the `Kernel` struct stored with an explicit `'static` lifetime. That's reasonable since it's probably a reasonable choice to statically allocate it anyway (we want these structures to tax sections of the binary such that an overflow will be caught by the linker, so not on the stack), but I wonder if this is a hard constraint, or if the lifetime can be parameterized.

In general, my intuition is that parameterizing lifetimes will give us more flexibility to eventually take out shared static variables (this is an intuition, I don't have a good argument for why that is).

I think it's fine to leave this as-is, if we convince ourselves that _clients_ of this interface (board `main.rs`s) will mostly likely still statically allocate, and in any case any relaxation of the lifetime constraint will still be compatible with static allocation (I think it is).

**Comment by alevy** [comment]
I added a CHANGELOG file with the changes I think matter for external projects. I'm approving with this addition, but leaving the final word on this to @bradjc. I'm also marking this P-Significant, which gives it another day for someone to raise the alarm bells.

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Since 1.2
+=========
+
+  * #1044 creates a `Kernel` struct with a method for the kernel's main loop,
```
Comment (ppannuto) [inline CHANGELOG.md]: It doesn't like like GH auto-links numbers in generic markdown files :(
https://github.com/tock/tock/blob/kernel-oop-just-work/CHANGELOG.md

**Comment by bradjc** [comment]
Looks good to me :)

**Pull request merged**